File: 39-combination-sum/39-combination-sum.java
import java.util.ArrayList;
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> parent = new ArrayList<>();
        csHelper(parent, new ArrayList<>(), candidates, target, 0, 0 );
        return parent;
    }
    
    public void csHelper(List<List<Integer>> parent, ArrayList<Integer> child,int[] nums, int target, int sum, int idx) {
        if(sum > target || idx == nums.length) return;
        else if(sum == target) {
            parent.add(new ArrayList<>(child));
            return;
        }
        
        child.add(nums[idx]);
        csHelper(parent, child, nums, target, sum + nums[idx], idx);
        child.remove(child.size()-1);
        csHelper(parent, child, nums, target, sum, idx + 1);
    }
}

File: 39-combination-sum/39-combination-sum.cpp
class Solution {
public:
    
    vector<vector<int>> output;
    
    vector<vector<int>> combinationSum(vector<int>& candidates, int target) {
        sort(candidates.begin(),candidates.end());
        vector<int> nums;
        dfs(candidates,target,0,nums,0);
        return output;
    }
    
    void dfs(vector<int>& candidates,int target,int sum,vector<int>& nums,int i){
        if(sum==target){
            output.push_back(nums);
            return;
        }
        if(sum>target || i>=candidates.size()){
            return;
        }
        nums.push_back(candidates[i]);
        dfs(candidates,target,sum+candidates[i],nums,i);
        
        nums.pop_back();
        dfs(candidates,target,sum,nums,i+1);
        
    }
};